### PR TITLE
feat(device, scheduler, capacity): implement CSI methods for volume operations

### PR DESCRIPTION
**Pull request by akhilerm**
- implement Create/Delete Volume and GetCapacity calls
- add DeviceNode CRD
- update vendor dependencies
- add VolumeWeighted and CapacityWeighted scheduler

### CODE REVIEW

File: pkg/device/device-util.go [hunk 1]
```diff
@@ -211,8 +211,8 @@ func RunCommand(cList []string) (string, error) {
 }
 
 // GetDiskList Todo
-func GetDiskList() ([]string, error) {
-	var result []string
+func GetDiskList() ([]apis.Device, error) {
```
Comment (akhilerm) [inline pkg/device/device-util.go]: @praveengt Have changed this method signature as we have a struct for the `Device` now which is introduced as part of the capacity tracking changes.

File: pkg/device/volume.go [hunk 2]
```diff
@@ -147,3 +153,49 @@ func RemoveVolFinalizer(vol *apis.DeviceVolume) error {
...
+
+// WaitForDeviceVolumeProcessed waits till the lvm volume becomes
+// ready or failed (i.e reaches to terminal state).
+func WaitForDeviceVolumeProcessed(ctx context.Context, volumeID string) (*apis.DeviceVolume, error) {
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, status.FromContextError(ctx.Err())
```
Comment (akhilerm) [inline pkg/device/volume.go]: NOTE: this code was copied over from [lvm-localpv](https://github.com/openebs/lvm-localpv/blob/master/pkg/lvm/volume.go#L115)

@iyashu This method needs to be removed as it was added [accidentally](https://github.com/grpc/grpc-go/pull/3561) in grpc as per [this release notes](https://github.com/grpc/grpc-go/releases/tag/v1.29.1).  

Didnt get the reason why, the method was used here, If I understand it better, I could replace it with the exact error that we intend to return to the caller.

**Review by akhilerm — COMMENTED** [review]
@praveengt @iyashu Have some doubts at some blocks of code. Can you please take a look. cc: @pawanpraka1

File: pkg/device/volume.go (hunk 2 above)
Comment (akhilerm) [reply pkg/device/volume.go]: The build will pass only once this error is fixed.

File: pkg/device/device-util.go [hunk 3]
```diff
@@ -81,22 +81,22 @@ func CreatePart(disk string, start int, partitionName string, size int) error {
 
 // GetAllPartsUsed Todo
 func getAllPartsUsed(diskName string, partitionName string) ([]partUsed, error) {
-	list, err := GetDiskList()
+	diskList, err := GetDiskList()
```
Comment (akhilerm) [inline pkg/device/device-util.go]: @praveengt PTAL.

File: pkg/device/device-util.go [hunk 4]
```diff
@@ -179,20 +179,20 @@ func GetVolumeDevPath(vol *apis.DeviceVolume) (string, error) {
 	//func GetPath(diskName string, partitionName string) (string, error) {
 	diskName := vol.Spec.DevName
 	partitionName := vol.Name
-	list, err := GetDiskList()
+	diskList, err := GetDiskList()
```
Comment (akhilerm) [inline pkg/device/device-util.go]: cc: @praveengt PTAL.

File: pkg/device/device-util.go [hunk 5]
```diff
@@ -248,14 +250,14 @@ func GetDiskList() ([]string, error) {
 
 // getAllPartsFree Todo
 func getAllPartsFree(diskName string) ([]partFree, error) {
-	list, err := GetDiskList()
+	diskList, err := GetDiskList()
```
Comment (akhilerm) [inline pkg/device/device-util.go]: cc: @praveengt

File: pkg/device/volume.go (hunk 2 above)
Comment (akhilerm) [reply pkg/device/volume.go]: Done. Ref: https://github.com/openebs/lvm-localpv/pull/37

File: pkg/apis/openebs.io/device/v1alpha1/devicenode.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	UUID string `json:"uuid"`
+
+	// Size specifies the total size of the device.
+	// +kubebuilder:validation:Required
+	Size resource.Quantity `json:"size"`
+	// Free specifies the available capacity of the device.
+	// +kubebuilder:validation:Required
+	Free resource.Quantity `json:"free"`
```
Comment (pawanpraka1) [inline pkg/apis/openebs.io/device/v1alpha1/devicenode.go:58]: does this make sense for device?

File: pkg/driver/controller.go [hunk 7]
```diff
@@ -78,6 +105,167 @@ func getRoundedCapacity(size int64) int64 {
...
+			vol, reschedule, err = waitForDeviceVolume(ctx, vol)
+			// If the device volume becomes ready or we can't reschedule failed volume,
+			// return the err.
+			if err == nil || !reschedule {
+				return vol, err
+			}
+		}
+	}
+
+	nmap, err := getNodeMap(params.Scheduler, params.DeviceName)
```
Comment (pawanpraka1) [inline pkg/driver/controller.go:236]: this is not needed for device localpv.

Comment (pawanpraka1) [reply pkg/driver/controller.go:236]: as we are going to support waitforfirstconsumer only, but we can keep this is plan is to support immediate mode.

File: pkg/device/device-util.go [hunk 8]
```diff
@@ -225,7 +225,9 @@ func GetDiskList() ([]string, error) {
 			continue
 		}
 		if tmp[5] == "disk" {
-			result = append(result, tmp[0])
+			result = append(result, apis.Device{
```
Comment (pawanpraka1) [inline pkg/device/device-util.go]: are we not fetching the size information?

File: pkg/mgmt/devicenode/devicenode.go [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+	"k8s.io/klog"
+
+	apis "github.com/openebs/device-localpv/pkg/apis/openebs.io/device/v1alpha1"
+	"github.com/openebs/device-localpv/pkg/builder/nodebuilder"
+	"github.com/openebs/device-localpv/pkg/device"
+	"github.com/openebs/device-localpv/pkg/equality"
+)
+
+func (c *NodeController) listDeviceNames() ([]apis.Device, error) {
+	return device.GetDiskList()
```
Comment (pawanpraka1) [inline pkg/mgmt/devicenode/devicenode.go]: I don't see GetDisklist return disk size info? how GetCapacity can use this if that is not populated?

**Comment by akhilerm** [comment]
cc: @AmitKumarDas

File: pkg/apis/openebs.io/device/v1alpha1/devicenode.go (hunk 6 above)
Comment (akhilerm) [reply pkg/apis/openebs.io/device/v1alpha1/devicenode.go:58]: Yes. We are using it.

File: deploy/device-operator.yaml [hunk 10]
```diff
@@ -258,10 +372,19 @@ spec:
             - "--feature-gates=Topology=true"
             - "--strict-topology"
             - "--leader-election"
+            - "--enable-capacity=true"
             - "--extra-create-metadata=true"
           env:
             - name: ADDRESS
               value: /var/lib/csi/sockets/pluginproxy/csi.sock
+            - name: POD_NAMESPACE
```
Comment (pawanpraka1) [inline deploy/device-operator.yaml:380]: why this is needed?

File: deploy/device-operator.yaml [hunk 11]
```diff
@@ -258,10 +372,19 @@ spec:
...
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.namespace
+            - name: POD_NAME
```
Comment (pawanpraka1) [inline deploy/device-operator.yaml:384]: and this?

File: pkg/driver/controller.go [hunk 12]
```diff
@@ -254,7 +503,97 @@ func (cs *controller) GetCapacity(
...
+	nodeNames, err := cs.filterNodesByTopology(segments)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	deviceNodeCache := cs.deviceNodeInformer.GetIndexer()
+	params := req.GetParameters()
+	deviceParam := helpers.GetInsensitiveParameter(&params, "devname")
+
+	klog.Infof("############## %+v", params)
```
Comment (pawanpraka1) [inline pkg/driver/controller.go:519]: debug?

File: pkg/driver/controller.go [hunk 13]
```diff
@@ -254,7 +503,97 @@ func (cs *controller) GetCapacity(
...
+		}
+		if !exists {
+			continue
+		}
+		deviceNode := v.(*apis.DeviceNode)
+		// rather than summing all free capacity, we are calculating maximum
+		// partition size that gets fit in given device.
+		// See https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1472-storage-capacity-tracking#available-capacity-vs-maximum-volume-size &
+		// https://github.com/container-storage-interface/spec/issues/432 for more details
+		klog.Infof("############## %+v", deviceNode)
```
Comment (pawanpraka1) [inline pkg/driver/controller.go:535]: debug?

**Pull request merged**